### DISCUSSION TITLE
Default behavior of environment variables and the setup of ${{ env.* }}

### DISCUSSION THREAD

**Discussion by rbellamy** [community/community#26468] [Actions]
<p>Up until now, I always assumed that whatever environment the runner was running in would be populated into the jobs.&lt;job id&gt;.env but I’ve just run some tests, and this does not seem to be the case.</p>
<p>The GitHub hosted runners come with a slew of environment variables via the /etc/environment file. Of particular interest to me are the various JAVA_HOME* variables, which point to JDKs (8, 11, 13, etc), with the default JAVA_HOME pointing to 11 (e.g. <span>JAVA_HOME=/usr/lib/jvm/zulu-8-azure-amd64</span>). I expected the following code to work, but it doesn’t:</p>
<pre><code>- name: Build with Maven
  env:
    JAVA_HOME: ${{ env.JAVA_HOME_8 }}
  runs: mvn -DskipTests compile
</code></pre>
<p>There are two surprises here:</p>
<ol>
<li>${{ env.JAVA_HOME_8 }} is empty, which causes,</li>
<li>$JAVA_HOME to be UNSET, so now it’s empty too.</li>
</ol>
<p><strong>NOTE: $JAVA_HOME_8 IS SET IN THE ENVIRONMENT.</strong></p>
<p>You can see an example of what I’m talking about here: <a href="https://github.com/terradatum/test-workflows/actions" rel="nofollow noopener">https://github.com/terradatum/test-workflows/actions</a>.</p>
<p><strong>I’d like to forward a feature request that ${{ env.* }} be populated with the current environment at startup.</strong></p>

**Comment by Yanjingzhu** [comment]
<p>Hi <a class="mention" href="/rbellamy">@rbellamy</a> , </p>
<p>In GitHub hosted ubuntu runner, there is not a system environment variable  <strong>JAVA_HOME_8.</strong></p>
<p>And in default, system environment variable is not included in env context, you could use <a href="https://github.com/actions/toolkit/blob/master/docs/commands.md#set-an-environment-variable" rel="nofollow noopener">set-env command</a> to set some system environment variables value to job envs . And then you could use it with  <strong>${{ env.envname }} </strong> syntax in the next steps in the current job. </p>
<p>There is a previous ticket for a similar issue : <a href="https://github.community/t/how-to-use-system-environment-variable-in-cache-actions-path/17960/2">https://github.community/t5/GitHub-Actions/How-to-use-system-environment-variable-in-cache-actions-path/m-p/51539#M8229</a> </p>

**Comment by rbellamy** [comment]
<p>Yes, in GitHub hosted runner there  <strong>IS</strong> a $JAVA_HOME_8 environment variable.</p>
<p>Please look at <a href="https://github.com/terradatum/test-workflows/runs/613285323?check_suite_focus=true" rel="nofollow noopener">https://github.com/terradatum/test-workflows/runs/613285323?check_suite_focus=true</a> and check the “Debug Environment” step to see the runtime environment of a GitHub Hosted runner.</p>

**Comment by rbellamy** [comment]
<p>And for a view of all tools installed on a GitHub Hosted runner: <a href="https://github.com/actions/virtual-environments/blob/master/images/linux/Ubuntu1804-README.md" rel="nofollow noopener">https://github.com/actions/virtual-environments/blob/master/images/linux/Ubuntu1804-README.md</a></p>
<p>Many of the provisioning scripts dump their runtime paths and environment variables into the “/etc/environment” file, which is then available to ALL actions in a GitHub hosted runner.</p>

**Comment by rbellamy** [comment, marked as answer]
<p>Asked and answered: <a href="https://github.com/actions/runner/issues/455" rel="nofollow noopener">https://github.com/actions/runner/issues/455</a></p>
<p>Final Answer: it was a deliberate choice to NOT populate the ${{ env.* }} space with a runner’s environment.</p>
